File: М11_ClusterVMS/clustervms/cluster/identity.py
"""Lesson 2 — who am I? Never the allocation index. The scheduler delivers
the Node's own Variable through its template as environment; this reads it,
and can also read the Variable directly for the values a template did not
carry (the column key, the heartbeat).

    nodes/<node>          node, config, revision, cameras
    nodes/<node>/pg       password
    nodes/<node>/key      hex            (М10's column key — the debt paid)
    nodes/<node>/epoch    epoch
    nodes/<node>/heartbeat ts, epoch     (for node_failover_seconds)
    nodes/<node>/failover last, worst    (RTO, worst case)
"""
from __future__ import annotations

import os
from dataclasses import dataclass

from .variables import Variables


@dataclass
class Identity:
    node: str
    config_object: str          # "" when the directory has never seen this Node
    config_revision: int
    camera_ids: list[int]
    column_key: bytes | None

    @property
    def seen_before(self) -> bool:
        return bool(self.config_object)


def from_environment(vars_: Variables | None = None) -> Identity:
    node = os.environ.get("NODE_ID") or os.environ.get("NOMAD_JOB_NAME")
    if not node:
        raise RuntimeError("NODE_ID is not set: a Node must know what it is before it does anything")
    cfg = os.environ.get("CONFIG_OBJECT", "")
    rev = int(os.environ.get("CONFIG_REVISION", "0") or 0)
    cams = [int(c) for c in os.environ.get("CAMERA_IDS", "").split(",") if c.strip()]
    key = None
    if vars_ is not None:
        items, _ = vars_.get(f"nodes/{node}/key")
        if items and items.get("hex"):
            key = bytes.fromhex(items["hex"])
        # the template may be stale by one render; the Variable is authoritative
        items, _ = vars_.get(f"nodes/{node}")
        if items:
            cfg = items.get("config", cfg)
            rev = int(items.get("revision", rev) or 0)
            cams = [int(c) for c in items.get("cameras", "").split(",") if c.strip()] or cams
    return Identity(node, cfg, rev, cams, key)

File: М11_ClusterVMS/clustervms-go/cluster/identity.go
package cluster

// Lesson 2 — who am I? Never the allocation index. The scheduler delivers
// the Node's own Variable through its template as environment; this reads
// it, and can also read the Variable directly for the values a template did
// not carry (the column key).
//
//	nodes/<node>           node, config, revision, cameras
//	nodes/<node>/pg        password
//	nodes/<node>/key       hex            (М10's column key — the debt paid)
//	nodes/<node>/epoch     epoch
//	nodes/<node>/failover  last, worst    (RTO, worst case)
//
// The heartbeat is NOT here: small, frequent, never queried — it is an
// object (<node>/heartbeat in the object store), see apphost.go.

import (
	"encoding/hex"
	"errors"
	"os"
	"strconv"
	"strings"
)

type Identity struct {
	Node           string
	ConfigObject   string // "" when the directory has never seen this Node
	ConfigRevision int64
	CameraIDs      []int64
	ColumnKey      []byte
}

func (i Identity) SeenBefore() bool { return i.ConfigObject != "" }

func parseIDs(s string) []int64 {
	var out []int64
	for _, p := range strings.Split(s, ",") {
		if p = strings.TrimSpace(p); p != "" {
			out = append(out, atoi(p))
		}
	}
	return out
}

// FromEnvironment reads NODE_ID (or NOMAD_JOB_NAME), CONFIG_OBJECT,
// CONFIG_REVISION and CAMERA_IDS; with a Variables client it then prefers
// the Variable — the template may be stale by one render.
func FromEnvironment(v Variables) (Identity, error) {
	node := os.Getenv("NODE_ID")
	if node == "" {
		node = os.Getenv("NOMAD_JOB_NAME")
	}
	if node == "" {
		return Identity{}, errors.New("NODE_ID is not set: a Node must know what it is before it does anything")
	}
	id := Identity{Node: node, ConfigObject: os.Getenv("CONFIG_OBJECT"),
		ConfigRevision: atoi(os.Getenv("CONFIG_REVISION")), CameraIDs: parseIDs(os.Getenv("CAMERA_IDS"))}
	if v == nil {
		return id, nil
	}
	if items, _, err := v.Get("nodes/" + node + "/key"); err == nil && items["hex"] != "" {
		if k, err := hex.DecodeString(items["hex"]); err == nil {
			id.ColumnKey = k
		}
	}
	items, _, err := v.Get("nodes/" + node)
	if err != nil {
		return id, err
	}
	if items != nil {
		if c, ok := items["config"]; ok {
			id.ConfigObject = c
		}
		if r, ok := items["revision"]; ok {
			id.ConfigRevision, _ = strconv.ParseInt(r, 10, 64)
		}
		if cams := parseIDs(items["cameras"]); len(cams) > 0 {
			id.CameraIDs = cams
		}
	}
	return id, nil
}
